lib/chef/provider/package/powershell.rb

```diff
@@ -0,0 +1,136 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/provider/package"
+require "chef/resource/powershell_package"
+require "chef/mixin/powershell_out"
+
+class Chef
+  class Provider
+    class Package
+      class Powershell < Chef::Provider::Package
+        include Chef::Mixin::PowershellOut
+
+        provides :powershell_package, os: "windows"
+
+        def load_current_resource
+          @current_resource = Chef::Resource::PowershellPackage.new(new_resource.name)
+          current_resource.package_name(new_resource.package_name)
+          current_resource.version(build_current_versions)
+          current_resource
+        end
+
+        def define_resource_requirements
+          super
+          if powershell_out("$PSVersionTable.PSVersion.Major").stdout.strip().to_i < 5
+            raise "Minimum installed Powershell Version required is 5"
+          end
+          requirements.assert(:install) do |a|
+            a.assertion { candidates_exist_for_all_uninstalled? }
+            a.failure_message(Chef::Exceptions::Package, "No candidate version available for #{packages_missing_candidates.join(", ")}")
+            a.whyrun("Assuming a repository that offers #{packages_missing_candidates.join(", ")} would have been configured")
+          end
+        end
+
+        def candidate_version
+          @candidate_version ||= build_candidate_versions
+        end
+
+        def install_package(names, versions)
+          # Installs the package specified with the version passed else latest version will be installed
+          names.each_with_index do |name, index|
+            powershell_out("Install-Package #{name} -Force -ForceBootstrap -RequiredVersion #{versions[index]}", { :timeout => @new_resource.timeout })
+          end
+        end
+
+        def remove_package(names, versions)
+          # Removes the package, if no version is passed, all installed version  will be removed
+          names.each_with_index do |name, index|
+            if !versions.nil?
+              if versions[index] != nil
+                powershell_out( "Uninstall-Package #{name} -Force -RequiredVersion #{versions[index]}", { :timeout => @new_resource.timeout })
+              else
+                powershell_out( "Uninstall-Package #{name} -Force", { :timeout => @new_resource.timeout })
+              end
+            end
+            if new_resource.version.nil?
+              stdout = 0
+              while stdout != ""
+                stdout = powershell_out( "Uninstall-Package #{name} -Force", { :timeout => @new_resource.timeout }).stdout
+                if !stdout.empty?
```

Comment:
As per @mwrock suggestion that if no version is passed during remove action, then we will uninstall all the versions of the package installed on the system. Until we keep getting some value in stdout, it means some version is getting removed of the package, and when we get stdout empty it means no version is left to be uninstalled.